File: mobile/ios/Shared/CountryRoutes.swift
import Foundation

/// The bundled country route set, read from the packed resource that
/// scripts/generate_cn_geoip.py produces from registry delegation data.
///
/// Deliberately a function and not a cached property. The parsed set is a few
/// hundred kilobytes of prefixes, the packet-tunnel extension runs against a
/// fixed memory profile shared with the Go runtime, and it is needed exactly
/// once per connect — so it is built, handed to a RoutePlan, and dropped before
/// the core starts.
enum CountryRoutes {
    /// The resource shipped inside the packet-tunnel extension.
    static let chinaResource = "cn-direct"

    enum Failure: LocalizedError {
        case resourceMissing(String)
        case notARouteSet
        case unsupportedVersion(UInt8)
        case truncated(expected: Int, found: Int)

        var errorDescription: String? {
            switch self {
            case .resourceMissing(let name):
                return "The bundled route set \(name) is missing from this build."
            case .notARouteSet:
                return "The bundled route set does not carry the expected header."
            case .unsupportedVersion(let version):
                return "The bundled route set is format version \(version), which this build cannot read."
            case .truncated(let expected, let found):
                return "The bundled route set should be \(expected) bytes but is \(found)."
            }
        }
    }

    private static let magic: [UInt8] = [0x51, 0x51, 0x47, 0x4F] // "QQGO"
    private static let formatVersion: UInt8 = 1
    private static let headerSize = 16
    private static let ipv4EntrySize = 5
    private static let ipv6EntrySize = 17

    /// Prefixes the registry delegates to China, sorted and already collapsed
    /// by the generator so no coalescing is needed here.
    ///
    /// This is where the set is *allocated*, which the registry is explicit is
    /// not where it is used. The UI says so; nothing downstream should imply
    /// otherwise.
    static func chinaDirect(in bundle: Bundle = .main) throws -> [IPPrefix] {
        guard let url = bundle.url(forResource: chinaResource, withExtension: "bin") else {
            throw Failure.resourceMissing(chinaResource)
        }
        return try decode(Data(contentsOf: url, options: .mappedIfSafe))
    }

    /// The packed set as bytes, for handing to the core.
    ///
    /// The Go side reads the same file by the same rules and searches it in
    /// place; it needs the bytes rather than the prefixes because a GEOIP rule
    /// is consulted per flow and the parsed form is a quarter of a megabyte to
    /// keep resident. Returns nil rather than throwing: a build without the
    /// resource still runs every rule that is not GEOIP, and the caller records
    /// that it is missing.
    static func packedChinaSet(in bundle: Bundle = .main) -> Data? {
        guard let url = bundle.url(forResource: chinaResource, withExtension: "bin") else {
            return nil
        }
        return try? Data(contentsOf: url, options: .mappedIfSafe)
    }

    /// How many blocks the bundled set holds, without parsing it.
    ///
    /// The counts sit in the header, so a screen can say how heavy the toggle
    /// is — several thousand routes, not a handful — without building or
    /// holding the prefixes. A claim the UI cannot substantiate is worse than
    /// no claim, and this one costs sixteen bytes.
    static func blockCount(in bundle: Bundle = .main) throws -> Int {
        guard let url = bundle.url(forResource: chinaResource, withExtension: "bin") else {
            throw Failure.resourceMissing(chinaResource)
        }
        let mapped = try Data(contentsOf: url, options: .mappedIfSafe)
        let counted = try counts(in: [UInt8](mapped.prefix(headerSize)))
        return counted.ipv4 + counted.ipv6
    }

    /// Reads the packed form. The layout is fixed-width and little else: a
    /// 16-byte header, then 5 bytes per IPv4 block and 17 per IPv6 block, all
    /// big-endian. scripts/test_cn_geoip.py holds the other half of this
    /// contract.
    static func decode(_ data: Data) throws -> [IPPrefix] {
        let bytes = [UInt8](data)
        let (ipv4Count, ipv6Count) = try counts(in: bytes)
        let expected = headerSize + ipv4EntrySize * ipv4Count + ipv6EntrySize * ipv6Count
        guard bytes.count == expected else {
            throw Failure.truncated(expected: expected, found: bytes.count)
        }

        var prefixes: [IPPrefix] = []
        prefixes.reserveCapacity(ipv4Count + ipv6Count)
        var offset = headerSize
        for _ in 0..<ipv4Count {
            let address = readBigEndian32(bytes, at: offset)
            if let prefix = IPPrefix(ipv4: address, length: Int(bytes[offset + 4])) {
                prefixes.append(prefix)
            }
            offset += ipv4EntrySize
        }
        for _ in 0..<ipv6Count {
            let high = readBigEndian64(bytes, at: offset)
            let low = readBigEndian64(bytes, at: offset + 8)
            if let prefix = IPPrefix(ipv6: high, low: low, length: Int(bytes[offset + 16])) {
                prefixes.append(prefix)
            }
            offset += ipv6EntrySize
        }
        return prefixes
    }

    /// Validates the header and returns the two block counts it carries.
    private static func counts(in bytes: [UInt8]) throws -> (ipv4: Int, ipv6: Int) {
        guard bytes.count >= headerSize, Array(bytes[0..<4]) == magic else {
            throw Failure.notARouteSet
        }
        guard bytes[4] == formatVersion else { throw Failure.unsupportedVersion(bytes[4]) }
        return (Int(readBigEndian32(bytes, at: 8)), Int(readBigEndian32(bytes, at: 12)))
    }

    private static func readBigEndian32(_ bytes: [UInt8], at offset: Int) -> UInt32 {
        (UInt32(bytes[offset]) << 24)
            | (UInt32(bytes[offset + 1]) << 16)
            | (UInt32(bytes[offset + 2]) << 8)
            | UInt32(bytes[offset + 3])
    }

    private static func readBigEndian64(_ bytes: [UInt8], at offset: Int) -> UInt64 {
        var value: UInt64 = 0
        for index in 0..<8 { value = (value << 8) | UInt64(bytes[offset + index]) }
        return value
    }
}

File: mobile/android/app/src/main/java/io/github/bojieli/queqiao/CountryRoutes.java
package io.github.bojieli.queqiao;

import android.content.Context;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Reads the packed country route set the build copies from the iOS client, the
 * same decoder as CountryRoutes.swift: "QQGO", a version byte, two big-endian
 * counts at offsets 8 and 12, then 5-byte IPv4 and 17-byte IPv6 entries.
 */
final class CountryRoutes {
    static final String CHINA_ASSET = "cn-direct.bin";
    static final String CHINA_CODE = "CN";

    private static final byte[] MAGIC = {0x51, 0x51, 0x47, 0x4F};
    private static final int FORMAT_VERSION = 1;
    private static final int HEADER_SIZE = 16;
    private static final int IPV4_ENTRY_SIZE = 5;
    private static final int IPV6_ENTRY_SIZE = 17;

    private CountryRoutes() {
    }

    static byte[] packedChinaSet(Context context) throws IOException {
        try (InputStream stream = context.getAssets().open(CHINA_ASSET)) {
            ByteArrayOutputStream buffer = new ByteArrayOutputStream();
            byte[] chunk = new byte[8192];
            int read;
            while ((read = stream.read(chunk)) != -1) {
                buffer.write(chunk, 0, read);
            }
            return buffer.toByteArray();
        }
    }

    static int blockCount(byte[] packed) throws IOException {
        int[] counts = counts(packed);
        return counts[0] + counts[1];
    }

    static List<RoutePolicy.RouteSpec> decode(byte[] packed) throws IOException {
        int[] counts = counts(packed);
        int expected = HEADER_SIZE + IPV4_ENTRY_SIZE * counts[0] + IPV6_ENTRY_SIZE * counts[1];
        if (packed.length != expected) {
            throw new IOException("The bundled route set should be " + expected + " bytes but is " + packed.length);
        }
        List<RoutePolicy.RouteSpec> prefixes = new ArrayList<>(counts[0] + counts[1]);
        int offset = HEADER_SIZE;
        for (int index = 0; index < counts[0]; index++, offset += IPV4_ENTRY_SIZE) {
            addPrefix(prefixes, packed, offset, 4);
        }
        for (int index = 0; index < counts[1]; index++, offset += IPV6_ENTRY_SIZE) {
            addPrefix(prefixes, packed, offset, 16);
        }
        return prefixes;
    }

    private static void addPrefix(List<RoutePolicy.RouteSpec> prefixes, byte[] packed, int offset, int width) {
        int length = packed[offset + width] & 0xFF;
        if (length > width * 8) {
            return;
        }
        BigInteger network = new BigInteger(1, Arrays.copyOfRange(packed, offset, offset + width));
        prefixes.add(new RoutePolicy.RouteSpec(network, length, width * 8));
    }

    private static int[] counts(byte[] packed) throws IOException {
        if (packed.length < HEADER_SIZE || !Arrays.equals(Arrays.copyOfRange(packed, 0, 4), MAGIC)) {
            throw new IOException("The bundled route set does not carry the expected header");
        }
        if ((packed[4] & 0xFF) != FORMAT_VERSION) {
            throw new IOException("The bundled route set is format version " + (packed[4] & 0xFF)
                    + ", which this build cannot read");
        }
        return new int[] {readBigEndian32(packed, 8), readBigEndian32(packed, 12)};
    }

    private static int readBigEndian32(byte[] bytes, int offset) {
        return ((bytes[offset] & 0xFF) << 24)
                | ((bytes[offset + 1] & 0xFF) << 16)
                | ((bytes[offset + 2] & 0xFF) << 8)
                | (bytes[offset + 3] & 0xFF);
    }
}
